### ISSUE TITLE
Tutorial example has a problem to enter new item into an empty list

### ISSUE THREAD

**Issue by FumioNonaka** [angular/angular#22120]
## I'm submitting a...

<pre><code>
[x] Documentation issue or request
</code></pre>

## Current behavior

"[Angular Example - Tour of Heroes: Part 6](https://angular.io/generated/live-examples/toh-pt6/stackblitz.html)" in the tutorial of Angular 5 has a problem to enter an item into the heroines.component.

1. Delete all items from the list of the heroines.component.
2. Enter new item in the text field and press the [add] button.

Then listed item has neither id nor name.  Browser console indicates the error message like below:

> `body: {error: "Collection 'heroes' id type is non-numeric or unknown. Can only generate numeric ids."}`

## Expected behavior

New item can be entered even if the list is empty.  I recommend to revise the code as "[Angular 5 Example - To be able to enter new item to empty list](https://qiita.com/FumioNonaka/items/3594a7d4a033f88522f7)" shows.

## Minimal reproduction of the problem with instructions

See the "Current behavior" above.

## What is the motivation / use case for changing the behavior?

Confused when learning the tutorial sample.

## Environment

<pre><code>
Angular version: 5.2.4

Browser:
- [x] Chrome (desktop) version 64.0.3282.140
- [x] Firefox version 58.0.1
- [x] Safari (desktop) version 11.0.3
 
For Tooling issues:
- Platform: macOS Sierra 10.12.6
</code></pre>

**Comment by trotyl** [comment]
Nice catch, currently the `in-memory-db` would only auto-increase the `id` [when there's existing `id` property and they're numeric](https://github.com/angular/in-memory-web-api/blob/f7e216d626207f5a553aa710463b5d06b39b8c43/src/in-mem/backend.service.ts#L513-L517):

```typescript
protected isCollectionIdNumeric<T extends { id: any }>(collection: T[], collectionName: string): boolean {
  // collectionName not used now but override might maintain collection type information
  // so that it could know the type of the `id` even when the collection is empty.
  return !!(collection && collection[0]) && typeof collection[0].id === 'number';
}
```

Suggested fix: override the `genId` method as we know a `Hero` should always have an `id`:

```typescript
genId(heroes: Hero[]): number {
  return heroes.length > 0 ? Math.max(...heroes.map(hero => hero.id)) + 1 : 1;
}
```

Demo [here](https://stackblitz.com/edit/angular-8r8msa?file=src%2Fapp%2Fin-memory-data.service.ts).

**Comment by FumioNonaka** [comment]
Thank you for your suggestion.  I appended the solution to [my article](https://qiita.com/FumioNonaka/items/3594a7d4a033f88522f7#overriding-a-method-of-angular-in-memory-web-api-module).

**Comment by avi182** [comment]
How can i check which method is responsible for the id generating?

**Comment by jenniferfell** [comment]
docarea=tutorial

**Comment by robphoenix** [comment]
For anyone else with the same problem that finds this issue, the `genId()` method should be added into the `in-memory-data.service.ts` file, there's a gist here - https://gist.github.com/robphoenix/3767cf08772347e02c2434bcf3b84635

**Comment by chembu** [comment]
Fixed as part of PR #24445
